src/components/vega/WordCloudViz.js

```diff
@@ -0,0 +1,53 @@
+import React, { useState, useLayoutEffect } from "react";
+import vegaEmbed from "vega-embed";
+import spec from "../../assets/vega/wordCloud";
+import data from "../../assets/vega/wordcloud_viz_data";
+
+const WordCloudViz = (props) => {
+  const [view, setView] = useState(null);
+  const [rendered, setRendered] = useState(false);
+
+  useLayoutEffect(() => {
+    if (!rendered) {
+      vegaEmbed("#wordcloud", spec, {
+        mode: "vega",
+        actions: false,
+        renderer: "svg",
+      }).then((res) => {
+        try {
+          res.view
+            .insert("corpus", data)
+            .runAsync()
+            .then((v) => {
+              // console.log(view)
+              setView(v);
+              // update the global state with the current mouseover
+              v.addEventListener("mouseover", (name, value) => {
```

Comment:
I remember you saying that you don't have a ton of experience with React hooks. And that's fine! :^)

The thing to know about `use[Layout]Effect` is that it's important to clean up after yourself. So when you add an event listener outside of React, you need to remove it yourself. You can return a _teardown functions from `use[Layout]Effect` to do this.

So I might write something like

```ts
useLayoutEffect(() => {
  const teardownFns = [];
  try {
    ...
    const mouseoverListener = (...) => { ... };
    v.addEventListener("mouseover", mouseoverListener);
    teardownFns.push(() => v.removeEventListener("mouseover", mouseoverListener);
  } catch (...) {
     ...
  } finally {
    return () => {
      for (const teardownFn of teardownFns) teardownFn();
    };
  }
});
```

Reply:
Alternatively, you could store the vega instance in state and use one effect per event listener. This is probably the most idiomatic thing to do. The idea of hooks is to be super modular, reusable bits of logic.

```ts
const [vega, setVega] = React.useState(null);

// Create the Vega instance and save it using setVega
...

// One effect per event listener that you're adding
React.useEffect(() => {
  if (!vega) return;
  const listener = () => { ... };
  vega.addEventListener(...);
  return () => vega.removeEventListener(...);
}, [vega]);
```

--- 

If you wanna get SUPER fancy, write your own hook. This is what hooks were designed to enable. They're like "logic mixins" that you can share.

```ts
function useVegaListener(vega, event, listener) {
    React.useEffect(() => {
    if (!vega) return;
    vega.addEventListener(...);
    return () => vega.removeEventListener(...);
  }, [vega, event, listener]);
}

// in the component
useVegaListener(vega, "mouseover", React.useCallback(() => { ... }, []);
```

Reply:
Looking at the vega-embed documentation, views have a `finalize` method which tears down all of the listeners, so I've added that to the return, but otherwise kept things consolidated as the sequence of events is important